api/resources_portal/views/document_views.py

```diff
@@ -0,0 +1,330 @@
+    CompoundSearchFilterBackend,
+    DefaultOrderingFilterBackend,
+    FacetedSearchFilterBackend,
+    FilteringFilterBackend,
+    OrderingFilterBackend,
+)
+from django_elasticsearch_dsl_drf.pagination import LimitOffsetPagination as ESLimitOffsetPagination
+from django_elasticsearch_dsl_drf.viewsets import DocumentViewSet
+from drf_yasg import openapi
+from drf_yasg.utils import swagger_auto_schema
+from elasticsearch_dsl import TermsFacet
+from six import iteritems
+
+from resources_portal.models import Material, Organization, User
+from resources_portal.models.documents import MaterialDocument, OrganizationDocument, UserDocument
+
+
+# Puts material search data into a form that the client will accept
+class MaterialDocumentSerializer(serializers.Serializer):
+    id = serializers.IntegerField(read_only=True)
+    category = serializers.CharField(read_only=True)
+    title = serializers.CharField(read_only=True)
+    url = serializers.CharField(read_only=True)
+    pubmed_id = serializers.CharField(read_only=True)
+    additional_metadata = serializers.CharField(read_only=True)
+    created_at = serializers.DateField(read_only=True)
+    updated_at = serializers.DateField(read_only=True)
+    organism = serializers.CharField(read_only=True)
+    has_publication = serializers.CharField(read_only=True)
+    has_pre_print = serializers.CharField(read_only=True)
+
+    class Meta:
+        model = Material
+        fields = (
+            "id",
+            "category",
+            "title",
+            "url",
+            "pubmed_id",
+            "additional_metadata",
+            "created_at",
+            "updated_at",
+            "organism",
+            "has_publication",
+            "has_pre_print",
+        )
+        read_only_fields = ("id", "created_at", "updated_at")
+
+
+class FacetedSearchFilterBackendExtended(FacetedSearchFilterBackend):
+    def aggregate(self, request, queryset, view):
+        """Extends FacetedSearchFilterBackend to add additional metrics to each bucket
+        https://github.com/barseghyanartur/django-elasticsearch-dsl-drf/blob/master/src/django_elasticsearch_dsl_drf/filter_backends/faceted_search.py#L19
+
+        We have the downloadable sample accession codes indexed for each experiment.
+        The cardinality metric, returns the number of unique samples for each bucket.
+        However it's just an approximate
+        https://www.elastic.co/guide/en/elasticsearch/reference/current/search-aggregations-metrics-cardinality-aggregation.html#_counts_are_approximate
+        I used the highest possible precision threshold, but this might increase the amount
+        of memory used.
+        """
+        facets = self.construct_facets(request, view)
+        for field, facet in iteritems(facets):
+            agg = facet["facet"].get_aggregation()
+            queryset.aggs.bucket(field, agg).metric(
+                "total_samples", "cardinality", field="id", precision_threshold=40000,
```

Comment:
This looks like it was copy pasted from refinebio. What's this doing? Is it still appropriate for this project?

Reply:
Yes. This adds a section in the returned JSON that lists each of the filterable fields (category, organism, etc) and how many of each filterable element there are. This is required for the front-end.

Reply:
Good catch! @Philinphiladelphia this might not be needed if you modify the method `transform_es_facets` below. Elastic search should already return the number of materials that match the query.

Reply:
Good catch. Modified to remove this.